pkg/services/ngalert/store/loki_range_to_instant.go

```diff
@@ -19,57 +19,74 @@ func (t dsType) isLoki() bool {
 	return t.DS.Type == datasources.DS_LOKI
 }
 
-func canBeInstant(r *models.AlertRule) bool {
+// canBeInstant checks if any of the query nodes that are loki range queries can be migrated to instant queries.
+// If any are migratable, those indices are returned.
+func canBeInstant(r *models.AlertRule) ([]int, bool) {
 	if len(r.Data) < 2 {
-		return false
-	}
-	// First query part should be range query.
-	if r.Data[0].QueryType != "range" {
-		return false
-	}
-
-	var t dsType
-	// We can ignore the error here, the query just won't be optimized.
-	_ = json.Unmarshal(r.Data[0].Model, &t)
-
-	if !t.isLoki() {
-		return false
-	}
-
-	exprRaw := make(map[string]interface{})
-	if err := json.Unmarshal(r.Data[1].Model, &exprRaw); err != nil {
-		return false
-	}
-
-	// Second query part should be and expression.
-	if !expr.IsDataSource(r.Data[1].DatasourceUID) {
-		return false
+		return nil, false
 	}
+	var (
+		optimizableIndices []int
+		canBeOptimized     = false
+	)
+	// Loop over query nodes to find all Loki range queries.
+	for i := range r.Data {
+		if r.Data[i].QueryType != "range" {
+			continue
+		}
+		var t dsType
+		// We can ignore the error here, the query just won't be optimized.
+		_ = json.Unmarshal(r.Data[i].Model, &t)
 
-	// Second query part should be "last()"
-	if val, ok := exprRaw["reducer"].(string); !ok || val != "last" {
-		return false
+		if !t.isLoki() {
+			continue
+		}
+		var reduceIndex int
+		// Loop over all query nodes to find the reduce node.
+		for ii := range r.Data {
+			// Second query part should be and expression.
+			if !expr.IsDataSource(r.Data[ii].DatasourceUID) {
+				continue
+			}
+			exprRaw := make(map[string]interface{})
+			if err := json.Unmarshal(r.Data[ii].Model, &exprRaw); err != nil {
+				continue
+			}
+			// Second query part should be "last()"
+			if val, ok := exprRaw["reducer"].(string); !ok || val != "last" {
+				continue
+			}
+			// Second query part should use first query part as expression.
+			if ref, ok := exprRaw["expression"].(string); !ok || ref != r.Data[i].RefID {
+				continue
+			}
+			reduceIndex = ii
+			break
```

Comment:
I think there's a bug here (as well as the previous version). It's not enough that there's a single `last` reduction that references the range query. **All** reduce expressions referencing the range query need to be `last` (or unused in the alert condition but that is not as important).

For example, this rule using a `vector(1)` range query will break with this optimization:

![image](https://github.com/grafana/grafana/assets/8484471/7a42a437-7473-41bc-b95f-e29bd53051fe)

![image](https://github.com/grafana/grafana/assets/8484471/b332f6c2-646c-4e7b-865d-bc4b7e9434a5)

Reply:
This is where I think we should consider moving the optimization code to the expr package as we can do this optimization on the parsed execution graph rather than having to re-create the graph ourselves. This is non-blocking, but just follows from Matt's comment about nested-loops and caching.

Reply:
Good find, fixed that and added a test.